### PR TITLE
array literals uses typed arrays; fix a jsgen bug

### PR DESCRIPTION
**Pull request by ringabout**
Ref https://github.com/timotheecour/Nim/issues/557

```nim
when true:
  proc main()=
    var b: array[10, byte]
    var b2 = [byte(1), 2]
    doAssert (b, b2) == ([byte(0), 0, 0, 0, 0, 0, 0, 0, 0, 0], [byte(1), 2])
  main()
```

Before this PR:
```js
function main_369098753() {
  var b_369098754 = new Uint8Array(10);
  var b2_369098755 = [1, 2];
  if (!(HEX3DHEX3D_369098758({ Field0: nimCopy(null, b_369098754, NTI369098758), Field1: nimCopy(null, b2_369098755, NTI369098759) }, { Field0: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], Field1: [1, 2] }))) {
    failedAssertImpl_218103865(makeNimstrLit("play.nim(5, 14) `(b, b2) == ([byte(0), 0, 0, 0, 0, 0, 0, 0, 0, 0], [byte(1), 2])` "));
  }
}
main_369098753();
```

After

```js
function main_369098753() {
  var b_369098754 = new Uint8Array(10);
  var b2_369098755 = new Uint8Array([1, 2]);
  if (!(HEX3DHEX3D_369098758({ Field0: nimCopy(null, b_369098754, NTI369098758), Field1: nimCopy(null, b2_369098755, NTI369098759) }, { Field0: new Uint8Array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), Field1: new Uint8Array([1, 2]) }))) {
    failedAssertImpl_218103865(makeNimstrLit("play.nim(5, 14) `(b, b2) == ([byte(0), 0, 0, 0, 0, 0, 0, 0, 0, 0], [byte(1), 2])` "));
  }
}
main_369098753();

```

### CODE REVIEW

File: compiler/jsgen.nim [hunk 1]
```diff
@@ -2147,21 +2151,40 @@ proc genSetConstr(p: PProc, n: PNode, r: var TCompRes) =
...
-    gen(p, n[i], a)
-    if a.typ == etyBaseIndex:
-      r.res.addf("[$1, $2]", [a.address, a.res])
-    else:
-      if not needsNoCopy(p, n[i]):
-        let typ = n[i].typ.skipTypes(abstractInst)
-        useMagic(p, "nimCopy")
-        a.res = "nimCopy(null, $1, $2)" % [a.rdLoc, genTypeInfo(p, typ)]
+  ## Constructs array or sequence.
+  ## Nim array(except array of `int64`, `bool`, `emum` etc. kind) maps to JS typed arrays.
```
Comment (timotheecour) [inline compiler/jsgen.nim]: that doesn't seem accurate for `array[3, SomeObject]`
maybe:
```suggestion
  ## Nim array of uint8..uint32, int8..int32 maps to JS typed arrays.
```

**Comment by timotheecour** [comment]
this needs a test.

* add this to lib/std/private/jsutils.nim:
```nim
proc jsTypeof*(a: auto): cstring =
  asm """`result` = typeof(`a`)"""

proc jsConstructorName*(a: auto): cstring =
  asm """`result` = `a`.constructor.name"""
```
(that file exists in devel but somehow not in your PR, which looks like it hasn't been rebased in a while?)

* and then add this somewhere (find an appropriate existing test file): (feel free to add as needed)
```nim
import std/private/jsutils
proc main()=
  template fn(a): untyped = jsConstructorName(a)
  doAssert fn(array[2, int8].default) == "Int8Array"
  doAssert fn(array[2, uint8].default) == "Uint8Array"
  doAssert fn(array[2, byte].default) == "Uint8Array"
  # doAssert fn(array[2, char].default) == "Uint8Array" # xxx fails; bug?
  doAssert fn(array[2, uint64].default) == "Array"
    # pending https://github.com/nim-lang/RFCs/issues/187 maybe use `BigUint64Array`
  doAssert fn([1'u8]) == "Uint8Array"
  doAssert fn([1'u16]) == "Uint16Array"
  doAssert fn([byte(1)]) == "Uint8Array"

main()
```

File: compiler/jsgen.nim [hunk 2]
```diff
@@ -2043,7 +2043,11 @@ proc genMagic(p: PProc, n: PNode, r: var TCompRes) =
   of mNew, mNewFinalize: genNew(p, n)
```
Comment (timotheecour) [inline compiler/jsgen.nim:2071]: add changelog entry in `## Compiler changes` mentioning this change

File: compiler/jsgen.nim [hunk 3]
```diff
@@ -2043,7 +2043,11 @@ proc genMagic(p: PProc, n: PNode, r: var TCompRes) =
...
   of mChr: gen(p, n[1], r)
```
Comment (timotheecour) [inline compiler/jsgen.nim:2072]: optional (or future work):
add performance test in the recent `tests/benchmarks` directory (see its readme; its meat to only run locally for now) to see if this does result in performance improvement (in at least some test cases)

File: compiler/jsgen.nim (hunk 2 above)
Comment (ringabout) [reply compiler/jsgen.nim:2071]: though it should, but I think the internal of JS array is not documented?

File: compiler/jsgen.nim (hunk 3 above)
Comment (ringabout) [reply compiler/jsgen.nim:2072]: But if `tests/benchmarks` is not tested, that's not easy to maintain;

File: compiler/jsgen.nim (hunk 2 above)
Comment (timotheecour) [reply compiler/jsgen.nim:2071]: it's not but should, but this is still an improvement worth mentioning for whoever looks at changelog
(manual is a "living" doc)

File: compiler/jsgen.nim (hunk 3 above)
Comment (timotheecour) [reply compiler/jsgen.nim:2072]: we can discuss this further in https://github.com/timotheecour/Nim/issues/425; short answer:
* for now, you can at least have `action: compile` (as in tests/benchmarks/ttls.nim)
* in future, we can also test in CI by passing `--define:expectedDuration:numSeconds` flag in CI so the test runs "fast enough", which still allows running locally with more accurate (but longer) timing, and generate whatever statistics/plots across commit history to find regressions

in any case, this benchmark can be added in future work

Comment (ringabout) [reply compiler/jsgen.nim:2072]: I agree.

File: tests/js/ttypedarray.nim [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  doAssert fn(array[2, int8].default) == "Int8Array"
+  doAssert fn(array[2, uint8].default) == "Uint8Array"
+  doAssert fn(array[2, byte].default) == "Uint8Array"
+  # doAssert fn(array[2, char].default) == "Uint8Array" # xxx fails; bug?
+  doAssert fn(array[2, uint64].default) == "Array"
+    # pending https://github.com/nim-lang/RFCs/issues/187 maybe use `BigUint64Array`
+  doAssert fn([1'u8]) == "Uint8Array"
+  doAssert fn([1'u16]) == "Uint16Array"
+  doAssert fn([byte(1)]) == "Uint8Array"
+
```
Comment (timotheecour) [inline tests/js/ttypedarray.nim:20]:
```suggestion
  doAssert fn([1.0'f32]) == "Float32Array"
  doAssert fn(array[2, float32].default) == "Float32Array"
  doAssert fn(array[2, float].default) == "Float64Array"
  doAssert fn(array[2, float64].default) == "Float64Array"
  doAssert fn([1.0]) == "Float64Array"
  doAssert fn([1.0'f64]) == "Float64Array"
```

File: changelog.md [hunk 5]
```diff
@@ -140,6 +140,9 @@ with other backends. see #9125. Use `-d:nimLegacyJsRound` for previous behavior.
 - TLS: OSX now uses native TLS (`--tlsEmulation:off`), TLS now works with importcpp non-POD types,
   such types must use `.cppNonPod` and `--tlsEmulation:off`should be used.
 
+- Now array literals(JS backend) uses JS typed arrays, for example `[byte(1), 2, 3]` generates `new Uint8Array([1, 2, 3])`.
```
Comment (timotheecour) [inline changelog.md]:
```suggestion
- Now array literals(JS backend) uses JS typed arrays when the corresponding js typed array exists, for example `[byte(1), 2, 3]` generates `new Uint8Array([1, 2, 3])`.
```

File: compiler/jsgen.nim [hunk 6]
```diff
@@ -2043,7 +2043,11 @@ proc genMagic(p: PProc, n: PNode, r: var TCompRes) =
...
   of mArrToSeq:
+    # initializing typed arrays doesn't need copy
     if needsNoCopy(p, n[1]):
+      # we change the kind from tyArray to tySequence because 
+      # `genArrayConstr` needs it.
+      skipTypes(n[1].typ, abstractVarRange).kind = tySequence
```
Comment (timotheecour) [inline compiler/jsgen.nim]: this sounds fishy for several reasons:
* it modifies source `n` (and doesn't reset it)
* what if it's inside some `compiles` statement, would this side effect affect code?
* ~~looks like it violates how case statement discriminant fields should be set~~ (EDIT: doesn't look like kind is a discriminant of PType but other points remain)

instead, can't we somehow pass a flag to `gen` so it handles it appropriately?

Comment (ringabout) [reply compiler/jsgen.nim]: Yeah, it seems not to be case statement discriminant fields

**Comment by ringabout** [comment]
I find a bug
```nim
proc hello(): array[5, int] = discard
var x = @(hello())
x.add(2)
echo x
```
x_369098759[0].push(2);;
               ^

TypeError: x_369098759[0].push is not a function
```js

**Comment by timotheecour** [comment]
@xflywind needs rebase

**Review by timotheecour — APPROVED** [review]

**Comment by timotheecour** [comment]
I'm still pondering the significance of https://github.com/timotheecour/Nim/issues/567, IMO that benchmark i wrote is contrived/not representative but would still be nice to have at least 1 non-contribed benchmark which shows benefit of typed array (but in future work is fine)

## bugfix: more type safe even in presence of casts
this PR fixes the following bug; we should add a test either in this PR or in followup PRs:
```nim
when true:
  import jsconsole
  template fn(a) =
    echo a
    console.log a
    var x = 257.2
    var b = cast[uint8](x)
    echo b
    a[0] = b
    echo a
    console.log a

  block:
    var a: array[3, uint8]
    a = [2'u8,4,5]
    fn(a)
  block:
    var a: array[3, uint8] = [2'u8,4,5]
    fn(a)
```


before PR: bug
[2, 4, 5]
Uint8Array(3) [ 2, 4, 5 ]
257
[1, 4, 5]
Uint8Array(3) [ 1, 4, 5 ]
[2, 4, 5]
[ 2, 4, 5 ]
257
[257, 4, 5]
[ 257.2, 4, 5 ]


after PR:
[2, 4, 5]
Uint8Array(3) [ 2, 4, 5 ]
257
[1, 4, 5]
Uint8Array(3) [ 1, 4, 5 ]
[2, 4, 5]
Uint8Array(3) [ 2, 4, 5 ]
257
[1, 4, 5]
Uint8Array(3) [ 1, 4, 5 ]

**Comment by Araq** [comment]
Is this still a draft?

**Comment by ringabout** [comment]
As https://stackoverflow.com/questions/13328658/are-the-advantages-of-typed-arrays-in-javascript-is-that-they-work-the-same-or-s said
> This Chrome developer comment provides some guidance that worked as of June 2012:
Normal arrays can be as fast as typed arrays if you do a lot of sequential access. Random access outside the bounds of the array causes the array to grow.
Typed arrays are fast for access, but slow to be allocated. If you create temporary arrays frequently, avoid typed arrays. (Fixing this is possible, but it's low priority.)
Micro-benchmarks such as JSPerf are not reliable for real-world performance.

Typed arrays are fast for access, but slow to be allocated, and normal arrays can be as fast as typed arrays if you do a lot of sequential access. Random access outside the bounds of the array causes the array to grow.

This PR is ready to review.

**Pull request merged**